compiler/rustc_ast_lowering/src/expr.rs

```diff
@@ -84,12 +84,28 @@ impl<'hir> LoweringContext<'_, 'hir> {
                 ExprKind::Tup(elts) => hir::ExprKind::Tup(self.lower_exprs(elts)),
                 ExprKind::Call(f, args) => {
                     if e.attrs.get(0).map_or(false, |a| a.has_name(sym::rustc_box)) {
-                        if let [inner] = &args[..] && e.attrs.len() == 1 {
+                        if e.attrs.len() != 1 {
+                            let err = self.tcx.sess.emit_err(RustcBoxAttributeError {
+                                span: e.attrs[0].span,
+                                reason: errors::RustcBoxAttrReason::Attributes,
+                            });
+                            return hir::Expr {
+                                hir_id,
+                                kind: hir::ExprKind::Err(err),
+                                span: self.lower_span(e.span),
+                            };
+                        }
+                        if let [inner] = &args[..]
+                            && let ExprKind::Path(_, path) = &f.kind
+                            && let [.., boxx, new] = path.segments.as_ref()
+                            && boxx.ident.name == sym::Box
+                            && new.ident.name == sym::new
```

Comment:
This would allow _any_ `Box`, not just the standard library's.

As `Box` is already a lang item (`owned_box`), you should be able to restrict it to methods of the type without any additional lang items. Alternatively, you could make `Box::new` a lang item and check for that specifically, which would limit it to only the one method.

Reply:
Alternatively, you can get the `Box` lang item, get its `tcx.associated_items`, and then just use https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/assoc/struct.AssocItems.html#method.find_by_name_and_kind to grab the `new` item.

Reply:
Sure, that would work as well. I wasn't sure on what was preferred, as I know effort is taken to avoid hard-coding names (even if that name cannot ever change).

Reply:
> Alternatively, you could make `Box::new` a lang item and check for that specifically, which would limit it to only the one method.

Or a diagnostic item, which #104363 is adding for `Box::new`.

Reply:
I don't think it's the best idea to use diagnostic items for a hard error like this. The compiler should be able to operate totally fine with all of the diagnostic items `cfg`'d out.

Reply:
Calling `tcx.lang_items()` here seems to ICE deep in the query system

Reply:
Oh yeah, makes sense. Can't check for lang items when the crate is being lowered, since that's where we discover the lang items.

Do you want to possibly move this check into a later phase of the compiler? There should probably be a post-typeck pass that you can stick this into... Or maybe implement it as a Forbid lint?

Reply:
That sounds like a good idea, I'll try that out soon.

Reply:
@compiler-errors Yeah a lang item might actually better. While this PR is not affecting code generation, it changes logic that decides on the set of accepted programs, which the "language" basically is. In #97774 I proposed a lang item too, also for wanting to do the same thing that this PR is doing.

Reply:
Looking at various ways of approaching this
- Implementing as a lint: Not sure this works, since once we've lowered to HIR, the call expression and its target is gone (unless we add the call target to `ExprKind::Box`, which seems a bit much for a lint)
- Moving to a later phase: Is there some crate where it would be appropriate/possible to walk HIR and replace a Call expr with the box?
- Another thought is that this could be implemented as `#[rustc_box]` on _any_ expression creating an `hir::ExprKind::Box` with that expression.

The last approach seems the cleanest to me, rather than requiring it be applied to a function which isn't actually being used. Would there be any issues with that?

Reply:
Huh -- we also could do an expr macro... `rustc_box!(expr)`

Reply:
Potentially, but I was hoping to remove `ExprKind::Box` from the AST with #108471

Reply:
Hm... ideally we'd remove box syntax from the parser, but there's nothing necessarily wrong with keeping it around in AST, like we have with type ascription via a macro.

Reply:
Regarding the suggestion to move the error to later in the compiler, it *might* be possible to move the lowering to HIR->THIR lowering in `rustc_mir_build`, but I'm not sure one can emit errors at that point. It's certainly worth a try tho.

Basically one would add a test [here](https://github.com/rust-lang/rust/blob/c4e0cd966062ca67daed20775f4e8a60c28e57df/compiler/rustc_mir_build/src/thir/cx/expr.rs#L299) whether there are any attributes, and if there is `#[rustc_box]` then one would create a THIR `ExprKind::Box` as done down below in the same method. As an example for code that checks for attributes, you can look for the `#[custom_mir]` attribute where the attr check is [here](https://github.com/rust-lang/rust/blob/dcca6a375bd4eddb3deea7038ebf29d02af53b48/compiler/rustc_mir_build/src/build/mod.rs#L488).

Regarding `rustc_box!(expr)`, I'm not a fan of using the privileged access that builtin-macros have for such purposes, but it seems to be the status quo (for now, unless we get `builtin#`). I think that the proposals to replace placement new go in the direction of using syntax that is [very similar to function calls](https://github.com/rust-lang/rfcs/pull/2884) though.

Reply:
> Regarding the suggestion to move the error to later in the compiler, it might be possible to move the lowering to HIR->THIR lowering in rustc_mir_build, but I'm not sure one can emit errors at that point. It's certainly worth a try tho.

Yeah, we definitely have errors emitted at and past this point. Notably pattern exhaustiveness errors, but others too.